### PR TITLE
[mypy] Add/fix type annotations for similarity search in machine learning

### PR DESCRIPTION
**Pull request by SteveKimSR**
$ mypy --strict machine_learning/similarity_search.py --ignore-missing-imports
Success: no issues found in 1 source file

### **Describe your change:**

Related Issue: #4052 

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by poyea — APPROVED** [review]
Thank you for your pull request!🤩

File: machine_learning/similarity_search.py [hunk 1]
```diff
@@ -30,7 +31,9 @@ def euclidean(input_a: np.ndarray, input_b: np.ndarray) -> float:
     return math.sqrt(sum(pow(a - b, 2) for a, b in zip(input_a, input_b)))
 
 
-def similarity_search(dataset: np.ndarray, value_array: np.ndarray) -> list:
+def similarity_search(
+    dataset: np.ndarray, value_array: np.ndarray
+) -> List[List[float]]:
```
Comment (dhruvmanila) [inline machine_learning/similarity_search.py:36]: The return type hint does not seem to be correct as per the given examples:
```
[[[0], 0.0]]
[[[0, 0, 0], 1.0]]
[[[0, 0, 0], 0.0], [[0, 0, 0], 1.0]]
```
Maybe: 
```python
List[List[Union[List[float], float]]]
```
I would suggest changing the return data type but that is out of scope for this PR.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: machine_learning/similarity_search.py (hunk 1 above)
Comment (SteveKimSR) [reply machine_learning/similarity_search.py:36]: @dhruvmanila My mistake... you are right.
Have a question: I've understood return data will be List[List[List[float], float]].
Your type above means List[List[List[float]]] or List[List[float]], am I correct??

Comment (SteveKimSR) [reply machine_learning/similarity_search.py]: Done :)

**Review by dhruvmanila — APPROVED** [review]
Sorry for the wait, I've been a bit busy. Thank you for your contribution! LGTM! 🎉
